### ISSUE TITLE
[Feature request] Engine repo mono tool 

### ISSUE THREAD

**Issue by reidbaker** [flutter/flutter#132807]
The engine is a necessarily complicated product and can be difficult to work in for contributors that do not frequently live in the engine code. Personally I have to open a wiki page or two every time I want to jump in the engine code. 

The feature request is for a mono tool designed for contributor and CI to use. 

The tools should be: 
* Written in dart 
    - Maximize number of contributors that have high proficiency
    - Runs on every platform we support
* Live in the engine repo `script/tool/bin/engine.dart
    - Allows for single repo change
    - Easier to keep up to date 
    - Code reviewable 
* Can build all targets `dart run engine.dart build --all` 
    - At 6 platforms this can be annoying to figure out 
* Can lint all targets `dart run engine.dart lint --all` 
* Can run all unit/(integration?) tests 
    - Building is hard enough, understanding how to run all tests few humans know how to do
* Used by CI in pre submit
    - Ensures build/lint/test continues to work because if they fail presubmit fails 
    - Gives contributors with domain expertise the ability to enforce or exempt code as it makes sense for their area of expertise 

This is not a new concept for the Dash Team. Flutter tool offers a public implementation of something similar and the packages repo has a contributor focused version of the same principals. This feature request is asking we apply the same idea to engine.

**Comment by zanderso** [comment]
This looks great. I think the easiest approach would be to start out with what I've suggested here: https://github.com/flutter/engine/pull/43856#issuecomment-1685429753. In short, a script/program that locally interprets the engine_v2 build configurations that live at https://github.com/flutter/engine/tree/main/ci/builders. If written in Dart, that would cover 1, 2, 3, 5, and 7.

(4) and (6) will bump up the complexity of the tool a fair amount. Whereas the other features only require harvesting commands from the build configurations, (4) requires synthesizing new commands from whole cloth. (6) would require rewriting the CI recipes to use the in-repo tool, and would be a heavy lift. I'd suggest that we not do (6) due to the complexity, and losing the fine-grained integration with LUCI.

**Comment by reidbaker** [comment]
If we do not do number 6 then the tool can (and will) be broken by changes since it is not executed during presubmit. I can understand not wanting to migrate all build, lint and test at once but I do think that directionaly for the tool to be healthy we should run it as part of presubmit.

**Comment by zanderso** [comment]
The idea would be to keep the tool working by way of unit tests, and by having it interpret the same build configurations that CI is interpreting, without any additional logic/behavior that could break. I understand this is unsatisfying, but I don't think we would have resources or buy-in from the infra team for anything fancier than that.

**Comment by reidbaker** [comment]
Another example of someone writing utility code to try to help better manage the engine and sharing it on the wiki for visibility. It looks helpful. 
 https://github.com/flutter/flutter/wiki/Setting-up-the-Engine-development-environment#manage-goma-with-a-script-recommended

**Comment by matanlurey** [comment]
Yeah similar for IDE support: https://github.com/flutter/flutter/wiki/Setting-up-the-Engine-development-environment#vscode-with-cc-intellisense-cc

**Comment by CodeDoctorDE** [comment]
I think this can be closed then?
